source/mesh_deformation/fastscape.cc

```diff
@@ -0,0 +1,1237 @@
+                    if ( cell->face(face_no)->boundary_id() != relevant_boundary)
+                      continue;
+
+                    std::vector<Tensor<1,dim> > vel(face_corners.size());
+                    fe_face_values.reinit(cell, face_no);
+                    fe_face_values[this->introspection().extractors.velocities].get_function_values(this->get_solution(), vel);
+
+                    for (unsigned int corner = 0; corner < face_corners.size(); ++corner)
+                      {
+                        const Point<dim> vertex = fe_face_values.quadrature_point(corner);
+
+                        // Find what x point we're at. Add 1 or 2 depending on if ghost nodes are used.
+                        const double indx = 1+use_ghost+vertex(0)/dx;
+
+                        // If our x or y index isn't close to a whole number, then it's likely an artifact
+                        // from using an over-resolved quadrature rule, in that case ignore it.
+                        if (abs(indx - round(indx)) >= precision)
+                          continue;
+
+                        /*
+                         * If we're in 2D, we want to take the values and apply them to every row of X points.
+                         */
+                        if (dim == 2)
+                          {
+                            for (int ys=0; ys<ny; ys++)
+                              {
+                                /*
+                                 * Fastscape indexes from 1 to n, starting at X and Y = 0, and increases
+                                 * across the X row. At the end of the row, it jumps back to X = 0
+                                 * and up to the next X row in increasing Y direction. We track
+                                 * this to correctly place the variables later on.
+                                 * Nx*ys effectively tells us what row we are in
+                                 * and then indx tells us what position in that row.
+                                 */
+                                const double index = round(indx)+nx*ys;
+
+                                temporary_variables[0].push_back(vertex(dim-1) - grid_extent[dim-1].second);
+                                temporary_variables[1].push_back(index-1);
+
+                                for (unsigned int i=0; i<dim; ++i)
+                                  {
+                                    if (this->convert_output_to_years())
+                                      temporary_variables[i+2].push_back(vel[corner][i]*year_in_seconds);
+                                    else
+                                      temporary_variables[i+2].push_back(vel[corner][i]);
+                                  }
+                              }
+                          }
+                        else
+                          {
+                            //Because indy only gives us the row we're in, we don't need to add 2 for the ghost node.
```

Comment:
```suggestion
                            // Because indy only gives us the row we're in, we don't need to add 2 for the ghost node.
```
Please clarify.